tests/integration/Core/Image/ImageStorage/ImageStorageTest.php

```diff
@@ -0,0 +1,207 @@
+                    $this->filePathNormalizer,
+                ]
+            )
+            ->getMock();
+    }
+
+    public function testHasFieldData(): void
+    {
+        self::assertTrue($this->storage->hasFieldData());
+    }
+
+    private function getContext(): array
+    {
+        return ['identifier' => 'LegacyStorage'];
+    }
+
+    /**
+     * @dataProvider providerOfFieldData
+     *
+     * @throws \eZ\Publish\Core\Base\Exceptions\InvalidArgumentException
+     * @throws \eZ\Publish\Core\IO\Exception\InvalidBinaryFileIdException
+     */
+    public function testStoreFieldDataDuringCreate(VersionInfo $versionInfo, Field $field): void
+    {
+        $binaryFile = $this->runCommonStoreFieldDataMocks($field);
+
+        $this->redecorator
+            ->expects(self::exactly(3))
+            ->method('redecorateFromSource')
+            ->with($binaryFile->uri)
+            ->willReturn($binaryFile->uri);
+
+        $this->storage->storeFieldData($versionInfo, $field, $this->getContext());
+
+        self::assertEquals(1, $this->gateway->countImageReferences($binaryFile->uri));
+    }
+
+    /**
+     * @dataProvider providerOfFieldData
+     *
+     * @depends testStoreFieldDataDuringCreate
+     */
+    public function testStoreFieldDataDuringUpdate(VersionInfo $versionInfo, Field $field): void
+    {
+        $binaryFile = $this->runCommonStoreFieldDataMocks($field);
+
+        $this->redecorator
+            ->expects(self::exactly(2))
+            ->method('redecorateFromSource')
+            ->with($binaryFile->uri)
+            ->willReturn($binaryFile->uri);
+
+        $this->storage->storeFieldData($versionInfo, $field, $this->getContext());
+
+        self::assertEquals(1, $this->gateway->countImageReferences($binaryFile->uri));
+    }
+
+    private function runCommonStoreFieldDataMocks(Field $field): BinaryFile
+    {
+        $this->filePathNormalizer
+            ->expects(self::once())
+            ->method('normalizePath')
+            ->willReturn($targetPath = '1/8/6/232-eng-GB/' . $field->value->externalData['fileName']);
+
+        $this->ioService
+            ->expects(self::once())
+            ->method('newBinaryCreateStructFromLocalFile')
+            ->with($field->value->externalData['inputUri'])
+            ->willReturn($newBinaryFileCreateStruct = new BinaryFileCreateStruct());
+
+        $this->ioService
+            ->expects(self::once())
+            ->method('createBinaryFile')
+            ->with($newBinaryFileCreateStruct)
+            ->willReturn($binaryFile = new BinaryFile(
+                [
+                    'id' => $targetPath,
+                    'uri' => $targetPath,
+                ]
+            ));
+
+        $this->ioService
+            ->expects(self::once())
+            ->method('getMimeType')
+            ->with($binaryFile->id)
+            ->willReturn('image/jpeg');
+
+        return $binaryFile;
+    }
+
+    public function providerOfFieldData(): array
```

Comment:
new[1] fashion:
```suggestion
    public function providerOfFieldData(): iterable
```

[1] actually a few years old